### PR TITLE
API improvements

### PR DESCRIPTION
**Pull request by cmark**
* Change ValueSet.compose to `0..1` from `0..*`
* Minor Java API improvements, syntactic sugars, etc.
* Support dynamic API docs
* Extract branch aware test execution to test.commons.
* Expose Elasticsearch Base64 UUID generator
* Minor REST API abstractions to simplify development of new REST API routes

### CODE REVIEW

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/ControllerExceptionMapper.java [hunk 1]
```diff
@@ -50,11 +52,21 @@
 		if (Throwables.getRootCause(ex).getMessage().toLowerCase().contains("broken pipe")) {
 	        return null; // socket is closed, cannot return any response    
 	    } else {
-	    	LOG.trace("Exception during request processing", ex);
+	    	if (PlatformUtil.isDevVersion()) {
```
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/ControllerExceptionMapper.java:55]: If anyone is wondering (I sure was): https://stackoverflow.com/questions/2621701/setting-log-level-of-message-at-runtime-in-slf4j/7829534

Quick summary: there is no way to do this programmatically in SLF4j without a helper method.

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/id/IDs.java [hunk 2]
```diff
@@ -13,25 +13,24 @@
...
-	}
-
-	public String getId() {
-		return id;
-	}
-	
-	public void setId(String id) {
-		this.id = id;
+	/**
+	 * Generates a time-based UUID (similar to Flake IDs), which is preferred when generating an ID to be indexed into a Lucene index asprimary key.
```
Comment (apeteri) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/id/IDs.java]: ⚠️ typo

```suggestion
	 * Generates a time-based UUID (similar to Flake IDs), which is preferred when generating an ID to be indexed into a Lucene index as primary key. 
```

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/domain/TransactionContext.java [hunk 3]
```diff
@@ -44,8 +44,9 @@
 	 * Adds the given {@link Object} to this transaction context as a completely new object. 
 	 * 
 	 * @param obj - the object to persist and add to the repository
+	 * @return the identifier if the given object is an instanceof of {@link Revision} or <code>null</code> if the object is not a {@link Revision}.
```
Comment (apeteri) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/domain/TransactionContext.java:47]: Javadoc is half-true: the default implementation returns an identifier for `CodeSystemEntry` and `CodeSystemVersionEntry` instances as well.

Comment (apeteri) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/domain/TransactionContext.java:47]: (Also, didn't `add` support indexing non-revision controlled documents before? Did this functionality disappear with 7.x, or does it only exist in my mind?)

File: snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedResourceRequest.java [hunk 4]
```diff
@@ -15,31 +15,27 @@
  */
 package com.b2international.snowowl.snomed.core.rest.domain;
 
-import com.b2international.snowowl.core.terminology.ComponentCategory;
-
+import com.b2international.snowowl.core.rest.domain.ResourceRequest;
+import com.google.common.base.MoreObjects.ToStringHelper;
 
 /**
  * @since 1.0
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedResourceRequest.java:22]: 🤔

**Review by apeteri — CHANGES_REQUESTED** [review]

**Review by nagyo — APPROVED** [review]
LGTM!

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/domain/TransactionContext.java (hunk 3 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/domain/TransactionContext.java:47]: Fixing it directly on 7.x after merging PR. Thanks!
Yeah, I think it did not, the `StagingArea` accepts any model, but the `TransactionContext` is primarily for `RevisionDocument`s.

**Pull request merged**